### ISSUE TITLE
Deno Standard GUI

### ISSUE THREAD

**Issue by rsp** [denoland/deno#3234]
Has there been any discussion about having a standard way of doing GUI in Deno? I don't necessarily mean that it all has to be in the core or std modules, and I don't mean anything complicated or very high-level, just something that lets us open native windows, show dialogs and generally create GUI apps using just JS/TS.

In issue #1629 about WebGL from January (the only issue related to GUI that I've found) @ry [you said](https://github.com/denoland/deno/issues/1629#issuecomment-459056787) that you want to support WebGL in core eventually. If it is possible to open a native window and draw using WebGL then it will be possible to use it to create a GUI framework as a third party module, but thinking about it I started wondering about a Web way to do GUIs, which is really HTML. I'm not arguing that it is the best way but this is certainly the most Web-style way. And since one of the goals of Deno is to have a similar API for the things that people are used to in browsers, then it seems like something worth thinking about.

If Deno doesn't use a Web way of creating GUIs then soon it will have to reinvent HTML and CSS, plus some logic of dispatching the click events etc. and it will soon get quite complicated.

If, on the other hand, Deno is able to open a window with HTML even with very limited functionality (like only text and canvas, for example), then it will be relatively easier to not go too far away from what people are used to in the browsers.

Of course having something like Electron bundled in Deno might be a little overkill (or maybe not? any thoughts on that?), but having a possibility to write an Electron-like framework as a third-party module **without** producing a separate Deno binary to be able to do that, would be something to think about.

Running `deno run https://www.example.com/gui-example.ts` to start a GUI program, instead of `deno-electron-like-fork run https://www.example.com/gui-example.ts` would be something to consider.

I wonder if there has already been any discussion about it.

**Comment by kitsonk** [comment]
@ry and I recently talked about it, and we presented the intent at TSConf, but there was no issue for it.  There was some conversation on [Gitter recently](https://gitter.im/denolife/Lobby?at=5da6a56a870fa33a4dfea6fb) as well.

**Comment by sholladay** [comment]
An Electron-ish thing built with Deno would be a big deal to me. I'd put a lot of effort into switching my apps to it, since the security improvements alone would be quite substantial. I suspect the resource efficiency would be much better, too, if Deno's startup time is anything to go by.

**Comment by rsp** [comment]
@kitsonk Thanks, I didn't know about it. Having a full-blown browser bundled would be something absolutely perfect, if only the overhead is not prohibitively expensive - but since you are already considering it with Ryan then I guess that happily it isn't.

Will what you were presenting with Ryan at TSConf about this be available online to watch? Or maybe it is already?

It would be great if Servo could be used here. The tight coupling with SpiderMonkey seems like a problem, unless it would be used just as a completely separated frontend and communicate with the main Deno process with a WebSocket or something like that, maybe having Deno transpiling the frontend code from TS to JS on the fly.

It is actually pretty interesting how it could be done - from GUI windows being just lightweight browsers, getting all assets via some proxy or file protocol and communicating via WebSockets with the Deno process, to having some duality like Electron, to having a browser part using the same engine or even the same process and event loop as the main Deno code. I wonder what is your and @ry's idea of what would be optimal here.

**Comment by rsp** [comment]
@sholladay Yeah, I am currently writing an app that needs a GUI which I am doing it in Electron but it is something that could be written in Deno with the current API with no problems if I could only do GUI with Deno (and unfortunately it cannot be a web app). I am concerned exactly with security of what I am doing right now, that the app will have full access to user's data, which could be nicely restricted with Deno. Not to mention the transpilation steps etc.

**Comment by brainkim** [comment]
Having a built-in module which bundles a cross-platform webview implementation would be the killer-feature that would get me to start experimenting/migrating from node.js.

**Comment by jardicc** [comment]
I have some basic feature requests. Few native dialogs. Alert, prompt, confirm and file/folder picker/save dialogs. Because sometimes you have app running in the background it will throw an error but user can't notice it so you have to show some dialog about what is going on and in Node.js there is no easy way to do that.

**Comment by erlend-sh** [comment]
Ya’ll should have a chat with @nothingismagick and the other fine folks at Tauri: https://github.com/tauri-apps/tauri/

A lot of what has been described here, they’re already working on or have already thought a great deal about.

**Comment by rsp** [comment]
@erlend-sh Are you sure this project is alive? The [tauri.studio](http://www.tauri.studio/) domain (from https://github.com/tauri-apps/tauri.studio) redirects to a domain parking page and on [tauri-apps.org](https://tauri-apps.org/) (the documentation link in the readme) I get ERR_CONNECTION_TIMED_OUT:
> This site can’t be reached tauri-apps.org took too long to respond.

**Comment by nothingismagick** [comment]
Yes - its alive. We're preparing our site to launch this month.

![image](https://user-images.githubusercontent.com/35242872/70623904-63136200-1c1f-11ea-94d8-7903a5a49b21.png)

**Comment by nothingismagick** [comment]
@rsp - we've had so many moving parts that we couldn't justify publishing the docs about how to use tauri until we've got a stabilized API and build system. Feel free to drop by our discord server if you want some quick support about getting setup with it.

On a sidenote, we do plan to release a deno CLI and rust integration once deno is stable enough for that too.

**Comment by rsp** [comment]
@nothingismagick Thanks for the info. The project after reading the readme looks almost too good to be true. Really looking forward to reading the docs when they are available. It's great to hear that you plan releasing a Deno CLI as well.

**Comment by nothingismagick** [comment]
yeah - we just a comment from a svelte user today. they confirmed a final linux build size of ca 600KB.

**Comment by nothingismagick** [comment]
But I know we can break the 500KB boundary for sure. Here's the discord if you want to stop in and have a chat:
https://discord.gg/SpmNs4S

**Comment by kitsonk** [comment]
It would be great if we remembered that this is a Deno issue tracker. :smile:

**Comment by adam-cyclones** [comment]
I would very much like to port my desktop environment to Deno, watching.

**Comment by axetroy** [comment]
Here is another idea for building GUI.

Call Chrome with [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/)

Build a GUI based on Chrome

This should be widely available

After all, most people's computers have Chromium kernel browsers. Including `Chromium Edge`/`Chrome`

**Comment by maxharris9** [comment]
I'm working on http://github.com/maxharris9/layout, and I'd like to port it to Deno.

**Comment by vincenzopalazzo** [comment]
+1

**Comment by aabbccsmith** [comment]
> Here is another idea for building GUI.
> 
> Call Chrome with [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/)
> ...
> Node.js implement: https://github.com/GoogleChromeLabs/carlo/
> Golang implement: https://github.com/zserge/lorca
> 
> I think this is a way to build a GUI application with lowest cost.

Why is this so frowned upon? This would mean super small package sizes when bundled and it would also be easier to build rather than a full rendering engine etc, unlike Electron apps? I second this idea!

**Comment by niklasbuesing** [comment]
Especially since future Windows versions ship with a Chromium distribution (Chromium Edge) by default

**Comment by crowlKats** [comment]
Not everyone has a chromium browser. I think Webview is the right way to go

**Comment by nothingismagick** [comment]
Indeed, chromium is a no go for FLOSS in some situations (e.g. you will not be able to ship Chromium based apps on the PureOS app store).

**Comment by niklasbuesing** [comment]
> Not everyone has a chromium browser. I think Webview is the right way to go

On Windows systems that don't have the Edge Browser, like Windows 7 which is still pretty common, Webviews uses MSHTML/Internet Explorer, which is a pretty big nope

**Comment by aabbccsmith** [comment]
> chromium is a no go for FLOSS in some situations

I would say that a minority as such could just not have supported. It's not an necessity to support every single platform - I'd rather have a Deno base framework replacement for Electron and support windows and mac exclusively. Electron's primary demographic so to speak is Windows and Mac devices and they don't plan to support minority OS's. 

TLDR, I can't see how not supporting OS's used by very few people (when the primary audience would be Windows and Mac) would be a problem

**Comment by galkowskit** [comment]
>On Windows systems that don't have the Edge Browser, like Windows 7 which is still pretty common

Windows 7 is EOL since 14th January. Microsoft is not supporting it, there is no reason for third parties to do so.

**Comment by niklasbuesing** [comment]
> Windows 7 is EOL since 14th January. Microsoft is not supporting it, there is no reason for third parties to do so.

When you're writing commercial software you can't just not support an operating system just because it isn't supported anymore, as long as its userbase is big enough you have to support it.

**Comment by nothingismagick** [comment]
At https://github.com/tauri-apps we are building the new official rust bindings to https://github.com/zserge/webview - Zserge's work will become the source headers for a pantheon of [bindings projects to webviews](https://github.com/webview), such as potentially https://github.com/eliassjogreen/deno_webview

Although tauri at the moment is focussing on rust, we anticipate making our rust-based builder and API available to other bindings, like deno - and have in fact started work on making a deno based CLI. (Which was paused because deno was missing a lot of stuff we needed.)

If you want to stop over and have a chat you can find the semi-official webview channel here:
https://discord.gg/KHmTZdH

(Semi official, only because at tauri we're hosting it until the webview org proper gets kickstarted).

**Comment by guckstift** [comment]
Isn't WebView using WebKit on Linux and isn't WebKit that Browser that refuses to implement WebGL2 for over 3 years?

**Comment by khionu** [comment]
Tauri has been working on new bindings to make it so the backing engine can be replaced. We have a "maybe in the future" goal of trying to bundle Servo, as an alternative for portability.

More directly responding, there aren't any good OSS alternatives to WebkitGTK for Linux. Qt is the only real alternative, but Qt has its licensing concerns.

**Comment by Wojtziech** [comment]
Basic set of dialog boxes in scripting languages can be used with https://en.wikipedia.org/wiki/Zenity

**Comment by danbulant** [comment]
> I have some basic feature requests. Few native dialogs. Alert, prompt, confirm and file/folder picker/save dialogs. Because sometimes you have app running in the background it will throw an error but user can't notice it so you have to show some dialog about what is going on and in Node.js there is no easy way to do that.

@jardicc you can already do alerts with [https://github.com/divy-work/autopilot-deno](autopilot), but there's no way to determine which button (close or confirm) was used, yet. Author is still working on it, and it will probably have more info soon (hopefully)

**Comment by pierredewilde** [comment]
> The status here is that there are both technical and non-technical reasons why integrating deno as the JS runtime in Electron instead of node is not possible or wanted at this time.
> https://github.com/electron/electron/issues/23613
> *This conversation has been locked as resolved...*
> 

Integrate Deno in Electron is a dream.
Keep Node in Electron would be a nightmare.

**Comment by awkj** [comment]
> > The status here is that there are both technical and non-technical reasons why integrating deno as the JS runtime in Electron instead of node is not possible or wanted at this time.
> > [electron/electron#23613](https://github.com/electron/electron/issues/23613)
> > _This conversation has been locked as resolved..._
> 
> Integrate Deno in Electron is a dream.
> Keep Node in Electron would be a nightmare.

It’s been almost 5 years since esm was released, and electron is still not supported

**Comment by gormonn** [comment]
Sorry, but it seems to me that HTML + CSS is a bad way to go.

I think HTML is a bad habit, a legacy from the previous generation when pages were static. Yes, now we have learned to revive them. However, due to the static nature of HTML, we were forced to look for "workarounds": virtual home, redraw control, etc.

We're trying to make interactive what was originally intended to be static, you know?

WebGL and Canvas are not about static. We draw frames every n-milliseconds. This is a completely different nature of the GUI, and different approaches.

**Comment by MierenManz** [comment]
To be honest. Alot of parts of a gui can be static.
So I think that we could use html and css for this

**Comment by stefnotch** [comment]
@gormonn Not to go too off-topic here, but that's overlooking a number of important aspects. Starting with
- WebGL and HTML aren't directly comparable, you'd have to compare a WebGL based GUI framework to HTML
  - Technically, modern browsers already use GPU acceleration to render trees of controls
- HTML is very accessible, as in, you can copy text, use screen readers, have descriptions of images, etc.
- HTML is still the de facto standard in browsers, making it hard for deno to suddenly invent a new GUI framework

**Comment by gormonn** [comment]
@stefnotch You are right, HTML is a lot, but it is not only pages on the Internet that must be indexed by search engines, read by various devices, including screen readers.

This is more about truly interactive desktop apps, kiosk apps, virtual guides, maps, billboards, media editors, storefronts, and more.

But you're right, this is offtopic.

**Comment by gormonn** [comment]
1. As for the Electron, it is very unstable. Sometimes it upsets me how new bugs appear when new major versions are released.
2. There are various bugs in chrome itself that Google hasn't fixed in years, and Electron cannot fix them.
3. Electron components are incompatible in terms of safety:
  - on the one hand NodeJS, with full access to the system
  - on the other hand, Chromium with its security settings that are focused on the user's safety when visiting web pages.
4. This creates problems with the design of the application. In addition, when packaging an application in AppImage (with / without ASAR), this also creates uncertain difficulties.

Moreover, I don't think you can just bind Electron to Deno. Unfortunately, I have no way to explain my concerns.

**Comment by KaKi87** [comment]
Here's a simpler and more obvious reason why a Deno GUI should be powered by HTML/CSS/JS : we are **web** devlopers.
That's it ^^
Thanks

**Comment by Soremwar** [comment]
@KaKi87 JavaScript is a general purpose programming language. Either way even if we say it isn't, WEBGL and WEBGPU are web technologies that have nothing to do with HTML rendering

**Comment by KaKi87** [comment]
What makes my web developer experience, the best development experience I ever had, is that, the whole time I spent learning front-end web development, I was actually learning how to code on 3 different platforms : browser, server, and desktop.
I never learned WebGL, and my whole point is based on that fact : once I knew HTML, CSS and JS, I didn't had to learn anything else.
That's what makes Electron's popularity.

**Comment by gormonn** [comment]
I agree with this statement. If this were not the case, there would be no concepts like CSS-in-JS.
> @ KaKi87 JavaScript - это язык программирования общего назначения. В любом случае, даже если мы говорим, что это не так, WEBGL и WEBGPU - это веб-технологии, которые не имеют ничего общего с рендерингом HTML и обеспечивают (возможно) лучший опыт разработки.

**Comment by SimonMeskens** [comment]
Not that it matters, but I also think that HTML/CSS/Web Standards are amazing. No need to reinvent the wheel.

**Comment by MierenManz** [comment]
> Not that it matters, but I also think that HTML/CSS/Web Standards are amazing. No need to reinvent the wheel.

I think that it would be nice to have a simple gui class aswell like autohotkey v2 has it. You just create a new gui like it's oop. It doesn't have to look good. But just basics for creating a simple window

**Comment by KaKi87** [comment]
> just basics for creating a simple window

Honestly that wouldn't be enough for me : I need to build real apps, not macros.

**Comment by MierenManz** [comment]
> > just basics for creating a simple window
> 
> Honestly that wouldn't be enough for me : I need to build real apps, not macros.

That's why I said that we should include a simple way that has default assets basically and one like webview is right now. Fully html css and js

**Comment by eliassjogreen** [comment]
I think integrating a more robust version of my webview module should suffice because abstractions and inventing new gui technologies is not really in the scope of the deno core or std (for example std/http only implements a basic web server without high level abstractions like routing). The web technologies are well established and using the platforms default provided browser does not add that much size to the executable unlike bundling something like chromium. Once webgpu lands and support for windowing (perhaps by using an external module like [pane](https://github.com/denosaurs/pane) 😉) is added my hope is that one would be free to implement any new abstractions and ideas in their own modules, writing their own renderers, dsl, dom systems, etc.

**Comment by jardicc** [comment]
> I think integrating a more robust version of my webview module should suffice because abstractions and inventing new gui technologies is not really in the scope ... The web technologies are well established and using the platforms default provided browser...

How would you communicate between Deno and OS provided Webview? Would it be 2 separated contexts?
